File: Serenity.Script.Imports/Q/Q.ScriptData.cs
using jQueryApi;
using System;
using System.Collections.Generic;
using System.Html;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;

namespace Serenity
{
    public static partial class Q
    {
        [Imported, ScriptNamespace("Q"), ScriptName("ScriptData")]
        public static class ScriptData
        {
            [InlineCode("Q.ScriptData.bindToChange({name}, {regClass}, {onChange})")]
            public static void BindToChange(string name, string regClass, Action onChange)
            {
            }

            [InlineCode("Q.ScriptData.triggerChange({name})")]
            public static void TriggerChange(string name)
            {
            }

            [InlineCode("Q.ScriptData.unbindFromChange({regClass})")]
            public static void UnbindFromChange(string regClass)
            {
            }

            [InlineCode("Q.ScriptData.ensure({name})")]
            public static object Ensure(string name)
            {
                return null;
            }

            [IncludeGenericArguments(false)]
            [InlineCode("Q.ScriptData.ensureAsync({name})")]
            public static Promise<TData> EnsureAsync<TData>(string name)
            {
                return null;
            }

            [InlineCode("Q.ScriptData.reload({name})")]
            public static object Reload(string name)
            {
                return null;
            }

            [IncludeGenericArguments(false)]
            [InlineCode("Q.ScriptData.reloadAsync({name})")]
            public static Promise<TData> ReloadAsync<TData>(string name)
            {
                return null;
            }

            [InlineCode("Q.ScriptData.canLoad({name})")]
            public static bool CanLoad(string name)
            {
                return false;
            }

            [InlineCode("Q.ScriptData.setRegisteredScripts({scripts})")]
            public static void SetRegisteredScripts(JsDictionary<string, string> scripts)
            {
            }

            [InlineCode("Q.ScriptData.set({name}, {value})")]
            public static void Set(string name, object value)
            {
            }
        }

        [IncludeGenericArguments(false)]
        public static TData GetRemoteData<TData>(string key)
        {
            return default(TData);
        }

        [IncludeGenericArguments(false)]
        public static Promise<TData> GetRemoteDataAsync<TData>(string key)
        {
            return null;
        }

        [IncludeGenericArguments(false)]
        public static Lookup<TItem> GetLookup<TItem>(string key)
        {
            return null;
        }

        [IncludeGenericArguments(false)]
        public static Promise<Lookup<TItem>> GetLookupAsync<TItem>(string key)
        {
            return null;
        }

        public static void ReloadLookup(string key)
        {
        }

        public static Promise ReloadLookupAsync(string key)
        {
            return null;
        }

        [IncludeGenericArguments(false)]
        public static List<PropertyItem> GetColumns(string key)
        {
            return null;
        }

        public static Promise<List<PropertyItem>> GetColumnsAsync(string key)
        {
            return null;
        }

        [IncludeGenericArguments(false)]
        public static List<PropertyItem> GetForm(string key)
        {
            return null;
        }

        public static Promise<List<PropertyItem>> GetFormAsync(string key)
        {
            return null;
        }

        public static string GetTemplate(string key)
        {
            return null;
        }

        public static Promise<string> GetTemplateAsync(string key)
        {
            return null;
        }

        public static bool CanLoadScriptData(string name)
        {
            return false;
        }
    }
}

File: Serenity.TypeScript.CoreLib/Q/Q.ScriptData.ts

namespace Q {

    export namespace ScriptData {
        let registered: { [key: string]: any } = {};
        let loadedData: { [key: string]: any } = {};

        export function bindToChange(name: string, regClass: string, onChange: () => void) {
            ($(document.body) as any).bind('scriptdatachange.' + regClass, function (e: any, s: string) {
                if (s == name) {
                    onChange();
                }
            });
        }

        export function triggerChange(name: string) {
            $(document.body).triggerHandler('scriptdatachange', [name]);
        }

        export function unbindFromChange(regClass: string) {
            $(document.body).unbind('scriptdatachange.' + regClass);
        }

        function syncLoadScript(url: string) {
            $.ajax({ async: false, cache: true, type: 'GET', url: url, data: null, dataType: 'script' });
        }

        function loadScriptAsync(url: string) {
            return Promise.resolve().then(function () {
                Q.blockUI(null);
                return Promise.resolve($.ajax({ async: true, cache: true, type: 'GET', url: url, data: null, dataType: 'script' }).always(function () {
                    Q.blockUndo();
                }));
            }, null);
        }

        function loadScriptData(name: string) {
            if (registered[name] == null) {
                throw new Error(Q.format('Script data {0} is not found in registered script list!', name));
            }

            name = name + '.js?' + registered[name];
            syncLoadScript(Q.resolveUrl('~/DynJS.axd/') + name);
        }

        function loadScriptDataAsync(name: string): PromiseLike<any> {
            return Promise.resolve().then(function () {
                if (registered[name] == null) {
                    throw new Error(Q.format('Script data {0} is not found in registered script list!', name));
                }

                name = name + '.js?' + registered[name];
                return loadScriptAsync(Q.resolveUrl('~/DynJS.axd/') + name);
            }, null);
        }

        export function ensure(name: string) {
            var data = loadedData[name];
            if (data == null) {
                loadScriptData(name);
            }

            data = loadedData[name];

            if (data == null)
                throw new Error(Q.format("Can't load script data: {0}!", name));

            return data;
        }

        export function ensureAsync(name: string): PromiseLike<any> {
            return Promise.resolve().then(function () {
                var data = loadedData[name];
                if (data != null) {
                    return Promise.resolve(data);
                }

                return loadScriptDataAsync(name).then(function () {
                    data = loadedData[name];
                    if (data == null) {
                        throw new Error(Q.format("Can't load script data: {0}!", name));
                    }
                    return data;
                }, null);
            }, null);
        }

        export function reload(name: string) {
            if (registered[name] == null) {
                throw new Error(Q.format('Script data {0} is not found in registered script list!', name));
            }
            registered[name] = (new Date()).getTime().toString();
            loadScriptData(name);
            var data = loadedData[name];
            return data;
        }

        export function reloadAsync(name: string) {
            return Promise.resolve().then(function () {
                if (registered[name] == null) {
                    throw new Error(Q.format('Script data {0} is not found in registered script list!', name));
                }
                registered[name] = (new Date()).getTime().toString();
                return loadScriptDataAsync(name).then(function () {
                    return loadedData[name];
                }, null);
            }, null);
        }

        export function canLoad(name: string) {
            return (loadedData[name] != null || registered[name] != null);
        }

        export function setRegisteredScripts(scripts: any[]) {
            registered = {};
            var t = new Date().getTime();
            for (var k in scripts) {
                registered[k] = scripts[k] || t;
            }
        }

        export function set(name: string, value: any) {
            loadedData[name] = value;
            triggerChange(name);
        }
    }

    export function getRemoteData(key: string) {
        return ScriptData.ensure('RemoteData.' + key);
    }

    export function getRemoteDataAsync(key: string) {
        return ScriptData.ensureAsync('RemoteData.' + key);
    }

    export function getLookup<TItem>(key: string): Lookup<TItem> {
        var name = 'Lookup.' + key;
        if (!ScriptData.canLoad(name)) {
            var message = 'No lookup with key "' + key + '" is registered. Please make sure you have a' +
                ' [LookupScript("' + key + '")] attribute in server side code on top of a row / custom lookup and ' +
                ' its key is exactly the same.';
            Q.notifyError(message);
            throw new Error(message);
        }
        return ScriptData.ensure('Lookup.' + key);
    }

    export function getLookupAsync<TItem>(key: string): PromiseLike<Lookup<TItem>> {
        return ScriptData.ensureAsync('Lookup.' + key);
    }

    export function reloadLookup(key: string) {
        ScriptData.reload('Lookup.' + key);
    }

    export function reloadLookupAsync(key: string) {
        return ScriptData.reloadAsync('Lookup.' + key);
    }

    export function getColumns(key: string) {
        return ScriptData.ensure('Columns.' + key);
    }

    export function getColumnsAsync(key: string) {
        return ScriptData.ensureAsync('Columns.' + key);
    }

    export function getForm(key: string) {
        return ScriptData.ensure('Form.' + key);
    }

    export function getFormAsync(key: string) {
        return ScriptData.ensureAsync('Form.' + key);
    }

    export function getTemplate(key: string) {
        return ScriptData.ensure('Template.' + key);
    }

    export function getTemplateAsync(key: string) {
        return ScriptData.ensureAsync('Template.' + key);
    }

    export function canLoadScriptData(name: string) {
        return ScriptData.canLoad(name);
    }
}
